### PR TITLE
Serf

### PR DESCRIPTION
**Pull request by sakateka**
subj
use Serf as membership provider

### CODE REVIEW

File: combainer/client.go [hunk 1]
```diff
@@ -195,9 +187,15 @@ func (cl *Client) Dispatch(parsingConfigName string, uniqueID string, shouldWait
 	defer cancelFunc()
 
 	cl.Log.WithFields(contextFields).Info("Start new iteration")
-	hosts, err := cl.Context.Hosts()
+	var hosts []string
+	serfMembers := cl.Context.Serf.Members()
```
Comment (noxiouz) [inline combainer/client.go]:
```go
hosts := make([]string, 0, len(serfMembers))
```

File: combainer/discovery.go [hunk 2]
```diff
@@ -142,9 +148,19 @@ func (s *SimpleFetcher) Fetch(groupname string) (hosts.Hosts, error) {
...
-			if put_err := s.Cache.Put(fetcherCacheNamespace, groupname, body); put_err != nil {
-				log.Infof("Put error: %s", put_err)
+			body, err = ioutil.ReadAll(resp.Body)
+			if err != nil {
+				log.Errorf("Failed to read response: %s", err)
+				body, err = s.Cache.Get(fetcherCacheNamespace, groupname)
+				if err != nil {
+					log.Errorf("Unable to read data from the cache: %s", err)
+				}
+				log.Errorf("%s answered with %s, but read response failed. Cache is used", url, resp.Status)
```
Comment (noxiouz) [inline combainer/discovery.go]: actualy cache read can fail. So cache is not used =)

**Comment by noxiouz** [comment]
LGTM

File: combainer/discovery.go (hunk 2 above)
Comment (sakateka) [inline combainer/discovery.go]: i lost return, fixed

File: combainer/client.go (hunk 1 above)
Comment (sakateka) [inline combainer/client.go]: yes, more clear :+1:

**Comment by sakateka** [comment]
@noxiouz I ready to merge

File: combainer/server/serf.go [hunk 3]
```diff
@@ -82,11 +82,11 @@ func (s *CombaineServer) setupSerf() (*serf.Serf, error) {
 
 	ips, err := net.LookupIP(conf.MemberlistConfig.Name)
 	if err != nil {
-		if len(ips) < 1 {
-			return nil, fmt.Errorf("failed to LookupIP for: %s", conf.MemberlistConfig.Name)
-		}
 		return nil, fmt.Errorf("failed to setup Serf: %s", err)
 	}
+	if len(ips) < 1 {
```
Comment (noxiouz) [inline combainer/server/serf.go]: len == 0?

**Pull request merged**